subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/NonHierarchicalFileWatcherUpdater.java

```diff
@@ -51,9 +51,9 @@
 
     private final WatchableHierarchies watchableHierarchies;
 
-    public NonHierarchicalFileWatcherUpdater(FileWatcher fileWatcher, Predicate<String> watchFilter) {
+    public NonHierarchicalFileWatcherUpdater(FileWatcher fileWatcher, Predicate<String> watchFilter, int maxHierarchiesToWatch) {
         this.fileWatcher = fileWatcher;
-        this.watchableHierarchies = new WatchableHierarchies(watchFilter);
+        this.watchableHierarchies = new WatchableHierarchies(watchFilter, maxHierarchiesToWatch);
```

Comment:
I wonder if there is a way to organize the code in a way that we don't need to pass this in here.

Reply:
As long as we want the non-hierarchical watchers respect the maximum number of hierarchies to watch, I don't think we can avoid passing the limit in here.